### PR TITLE
Creates an autocontrast KPL

### LINKED ISSUE #81: AutoContrast preprocessing layer
Reference implementation: https://github.com/tensorflow/similarity/blob/906d141b75acf00d0e1d37dd0d88c432bbb30f57/tensorflow_similarity/augmenters/img_augments.py#L542

### PR DESCRIPTION
**Pull request by LukeWood**
fixes #81

### CODE REVIEW

File: keras_cv/layers/preprocessing/auto_contrast.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+from keras_cv.utils import preprocessing
+
+
+class AutoContrast(tf.keras.__internal__.layers.BaseImageAugmentationLayer):
+    """Performs the AutoContrast operation on an image.
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/auto_contrast.py:21]: Can you add more details about the layer behavior here? eg the math part of how the pixel value is updated.

**Review by qlzh727 — CHANGES_REQUESTED** [review]
Thanks for the PR.

File: keras_cv/layers/preprocessing/auto_contrast.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            image = image * scale + offset
+            image = tf.clip_by_value(image, 0.0, 255.0)
+            return image
+
+        # we can't scale uniform channels
+        result = tf.cond(high > low, lambda: scale_values(image), lambda: image)
+        return result
+
+    def augment_image(self, image, transformation=None):
+        image = preprocessing.transform_value_range(
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/layers/preprocessing/auto_contrast.py:60]: https://github.com/keras-team/keras-cv/pull/198

File: keras_cv/layers/preprocessing/auto_contrast.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            image, original_range=self.value_range, target_range=(0, 255)
+        )
+        channels = tf.shape(image)[-1]
+        result = []
+        for c in tf.range(channels):
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/layers/preprocessing/auto_contrast.py]: Good point - done

File: keras_cv/layers/preprocessing/auto_contrast.py (hunk 1 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/auto_contrast.py:21]: Done

File: keras_cv/layers/preprocessing/auto_contrast_test.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+# limitations under the License.
+
+import tensorflow as tf
+
+from keras_cv.layers import preprocessing
+
+
+class AutoContrastTest(tf.test.TestCase):
+    def test_auto_contrast_expands_value_range(self):
+        img = tf.constant([0, 128], dtype=tf.float32)
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/layers/preprocessing/auto_contrast_test.py:35]: Done

File: keras_cv/layers/preprocessing/auto_contrast.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    def __init__(
+        self,
+        value_range=(0, 255),
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.value_range = value_range
+        # tf.vectorized_map can't handle iteration over a tf.range.
+        self.auto_vectorize = False
```
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/layers/preprocessing/auto_contrast.py]: Done

**Comment by LukeWood** [comment]
> Thanks for the PR.

Thanks - updated per comments

File: keras_cv/layers/preprocessing/auto_contrast.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        super().__init__(**kwargs)
+        self.value_range = value_range
+
+    def augment_image(self, image, transformation=None):
+        original_image = image
+        image = preprocessing.transform_value_range(
+            image, original_range=self.value_range, target_range=(0, 255)
+        )
+
+        low = tf.reduce_min(image, axis=0)
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/auto_contrast.py:50]: Shouldn't the low and high be the value for channel? eg the low should be:
` tf.reduce_min(tf.reduce_min(image, axis=0), axis=0)`

(It reduced both height and width dimension, and only keep for the channel)

File: keras_cv/layers/preprocessing/auto_contrast_test.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+        layer = preprocessing.AutoContrast(value_range=(0, 255))
+        ys = layer(img)
+
+        # Some pixels should be replaced with fill value
+        self.assertTrue(tf.math.reduce_any(ys[0] == 1.0))
+        self.assertTrue(tf.math.reduce_any(ys[0] == 1.0))
+
+    def test_auto_contrast_expands_value_range(self):
+        img = tf.constant([0, 128], dtype=tf.float32)
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/auto_contrast_test.py:35]: Can we have a test that have different value for each column/row and channel? eg
[[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]]]

**Review by qlzh727 — CHANGES_REQUESTED** [review]

File: keras_cv/layers/preprocessing/auto_contrast.py (hunk 6 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/auto_contrast.py]: ah yes - this was a mistake.  sorry about that.

File: keras_cv/layers/preprocessing/auto_contrast_test.py (hunk 7 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/auto_contrast_test.py:34]: Added - great idea.  Thanks

File: keras_cv/layers/preprocessing/auto_contrast_test.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    def test_auto_contrast_different_values_per_channel(self):
+        img = tf.constant(
+            [[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]]], dtype=tf.float32
+        )
+        img = tf.expand_dims(img, axis=0)
+
+        layer = preprocessing.AutoContrast(value_range=(0, 255))
+        ys = layer(img)
+
+        self.assertTrue(tf.math.reduce_any(ys[0, ..., 0] == 0.0))
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/auto_contrast_test.py:54]: Since the input value are fixed, can we assert the output as the fixed value as well?

eg, the result are:

[[[  0.        0.        0.     ]
  [ 85.       85.       85.     ]]

 [[170.    170.      170.     ]
  [255.  255.  255.  ]]], shape=(2, 2, 3), dtype=float32)

**Review by qlzh727 — APPROVED** [review]

File: keras_cv/layers/preprocessing/auto_contrast_test.py (hunk 8 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/auto_contrast_test.py:54]: Good point.  Thanks

File: keras_cv/layers/preprocessing/auto_contrast_test.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        self.assertTrue(tf.math.reduce_any(ys[0, ..., 0] == 255.0))
+        self.assertTrue(tf.math.reduce_any(ys[0, ..., 1] == 255.0))
+
+        self.assertAllClose(
+            ys,
+            [
+                [
+                    [
+                        [
+                            0.0,
```
Comment (qlzh727) [inline keras_cv/layers/preprocessing/auto_contrast_test.py:66]: err, we might want to make this more readable, and make the formatter to not forcefully convert it like this.
